src/state/require.rs

```diff
@@ -58,4 +58,9 @@ impl<'a, 'b, P> StateReq<'a, 'b, P> {
             .then_some(())
             .ok_or_else(StateError::required_missing::<Source, T>)
     }
+
+    /// Returns the `state` for advanced requirements.
+    pub fn state<'c>(&'c self) -> &'c State<'b, P> {
+        self.0
+    }
```

Comment:
What use case did you have in mind?

Reply:
I have a very awkward requirement for my master's thesis, and this was my first attempt to fix it, but the real issue lies deeper. 

I'm doing a kind of automated algorithm design with MAHF, so I have a hyper-heuristic that designs `Configuration`s and evaluates them on some problems. Each problem of course has an `Evaluator`, which is passed to the "meta" algorithm design problem. 

Now, the designed algorithm is actually itself composed of multiple metaheuristics, which each have their own `State` (I'm basically evolving the structure and composition of a heterogenous island model), but I only have one `Evaluator` (because it generally is not `Clone`), so I have to insert and remove the `Evaluator` from each `State` before executing the respective metaheuristic. 

This works well for the `init` and `execute` phases, but `require` was meant as a read-only phase, and this is not possible in my case. I'm just skipping the phase for now, but `require` should really also pass a `&mut StateReq` and allow for `State` modifications. These are just things that you really wouldn't think about when designing the `Component` interface.

On another note, is `Clone` a hard thing to support for e.g. the COCO evaluator? It holds a reference to the COCO problem, right, but it's not possible to instantiate COCO problems directly, only through their `Suite` or something?

Reply:
> On another note, is Clone a hard thing to support for e.g. the COCO evaluator? It holds a reference to the COCO problem, right, but it's not possible to instantiate COCO problems directly, only through their `Suite` or something?

Yes, `coco::Problem` instances can only be created from the Suite. That's also why they do not implement `Clone`. You can also only have one `Problem` per `Suite` at a time.

> This works well for the `init` and `execute` phases, but `require` was meant as a read-only phase, and this is not possible in my case. I'm just skipping the phase for now, but `require` should really also pass a `&mut StateReq` and allow for `State` modifications. These are just things that you really wouldn't think about when designing the `Component` interface.

So you would want to insert an evaluator in the `require` phase? Couldn't you insert a stub evaluator when you remove the real one? So during `init` and `execute` you use the real evaluator, and after they ran, you remove it and insert a dummy? That way, the `require` phase should pass, and you don't have to mutate the state during `require`.

Reply:
I guess that would be possible. A little bit hacky, though.

Reply:
So do you still need this function?

Reply:
No, let's remove it, then.